File: src/Application/Scene/SceneBuilder.ts
import { Scene } from '../../Core/Scene';
import { Mesh } from '../../Core/Geometry/Mesh';
import { Color } from '../../Core/Math/Color';
import { Vector3 } from '../../Core/Math/Vector3';
import { Light } from '../../Core/Light';

export class SceneBuilder {
    static createDemoScene(): Scene {
        const scene = new Scene();

        const floor = Mesh.createPlane(20, 20);
        floor.position = new Vector3(0, -1, 0);
        floor.color = Color.gray();
        scene.addMesh(floor);

        const cube1 = Mesh.createCube(1);
        cube1.position = new Vector3(-3, 0, 0);
        cube1.color = Color.red();
        scene.addMesh(cube1);

        const cube2 = Mesh.createCube(0.8);
        cube2.position = new Vector3(0, 0, 0);
        cube2.color = Color.green();
        scene.addMesh(cube2);

        const cube3 = Mesh.createCube(1.2);
        cube3.position = new Vector3(3, 0, 0);
        cube3.color = Color.blue();
        scene.addMesh(cube3);

        const sphere1 = Mesh.createSphere(0.7, 20, 20);
        sphere1.position = new Vector3(-2, 2, -2);
        sphere1.color = Color.yellow();
        scene.addMesh(sphere1);

        const sphere2 = Mesh.createSphere(0.6, 20, 20);
        sphere2.position = new Vector3(2, 2, -2);
        sphere2.color = Color.cyan();
        scene.addMesh(sphere2);

        const sphere3 = Mesh.createSphere(0.5, 20, 20);
        sphere3.position = new Vector3(0, 3, -1);
        sphere3.color = Color.magenta();
        scene.addMesh(sphere3);

        const backWall = Mesh.createCube(1);
        backWall.position = new Vector3(0, 0.5, -5);
        backWall.scale = new Vector3(10, 3, 0.5);
        backWall.color = new Color(0.3, 0.3, 0.4, 1);
        scene.addMesh(backWall);

        const pillar1 = Mesh.createCube(1);
        pillar1.position = new Vector3(-5, 1, -3);
        pillar1.scale = new Vector3(0.5, 4, 0.5);
        pillar1.color = new Color(0.6, 0.4, 0.2, 1);
        scene.addMesh(pillar1);

        const pillar2 = Mesh.createCube(1);
        pillar2.position = new Vector3(5, 1, -3);
        pillar2.scale = new Vector3(0.5, 4, 0.5);
        pillar2.color = new Color(0.6, 0.4, 0.2, 1);
        scene.addMesh(pillar2);

        scene.addLight(Light.createAmbient(Color.white(), 0.15));
        scene.addLight(Light.createDirectional(new Vector3(0.5, 0.7, 1.0).normalize(), Color.white(), 1.0));
        scene.addLight(Light.createPoint(new Vector3(0, 2, 0), new Color(1.0, 0.9, 0.8, 1.0), 1.5));

        return scene;
    }

    static createRotatingScene(): Scene {
        const scene = new Scene();

        const centerCube = Mesh.createCube(1);
        centerCube.position = Vector3.zero();
        centerCube.color = Color.white();
        scene.addMesh(centerCube);

        for (let i = 0; i < 8; i++) {
            const angle = (i / 8) * Math.PI * 2;
            const radius = 3;

            const sphere = Mesh.createSphere(0.4, 16, 16);
            sphere.position = new Vector3(
                Math.cos(angle) * radius,
                Math.sin(i * 0.5),
                Math.sin(angle) * radius
            );

            const hue = i / 8;
            sphere.color = new Color(
                Math.abs(Math.sin(hue * Math.PI * 2)),
                Math.abs(Math.sin((hue + 0.33) * Math.PI * 2)),
                Math.abs(Math.sin((hue + 0.66) * Math.PI * 2)),
                1
            );

            scene.addMesh(sphere);
        }

        scene.addLight(Light.createAmbient(Color.white(), 0.2));
        scene.addLight(Light.createDirectional(new Vector3(1, 1, 1).normalize(), Color.white(), 0.8));

        return scene;
    }
}

File: VulkanEngine/Application/Services/SceneBuilder.cs
using VulkanEngine.Core.Entities;
using VulkanEngine.Core.Math;

namespace VulkanEngine.Application.Services;

public class SceneBuilder
{
    public static Core.Scene CreateDemoScene()
    {
        var scene = new Core.Scene();

        scene.Camera.Position = new Vector3(0, 0, 3);
        scene.Camera.Target = new Vector3(0, 0, 0);

        scene.AddLight(Light.CreateDirectional(
            new Vector3(0.5f, 0.7f, 1.0f),
            Color.White,
            1.5f
        ));

        scene.AddLight(Light.CreateAmbient(
            Color.White,
            0.3f
        ));

        scene.AddTriangle(new Triangle(
            new Vector3(-2, -2, 0),
            new Vector3(2, -2, 0),
            new Vector3(0, 2, 0),
            Color.Red
        ));

        CreateCubeTriangles(new Vector3(-2, 0, 0), new Vector3(1, 1, 1), Color.Red, scene);
        CreateCubeTriangles(new Vector3(0, 0, 0), new Vector3(0.8f, 1, 0.8f), Color.Green, scene);
        CreateCubeTriangles(new Vector3(2, 0, 0), new Vector3(1.2f, 1.2f, 1.2f), Color.Blue, scene);

        CreateFloorPlane(new Vector3(0, -1, 0), 20, new Color(0.6f, 0.6f, 0.6f), scene);

        return scene;
    }

    private static void CreateCubeTriangles(Vector3 center, Vector3 size, Color color, Core.Scene scene)
    {
        var halfSize = size / 2f;

        var v0 = center + new Vector3(-halfSize.X, -halfSize.Y, -halfSize.Z);
        var v1 = center + new Vector3(halfSize.X, -halfSize.Y, -halfSize.Z);
        var v2 = center + new Vector3(halfSize.X, halfSize.Y, -halfSize.Z);
        var v3 = center + new Vector3(-halfSize.X, halfSize.Y, -halfSize.Z);
        var v4 = center + new Vector3(-halfSize.X, -halfSize.Y, halfSize.Z);
        var v5 = center + new Vector3(halfSize.X, -halfSize.Y, halfSize.Z);
        var v6 = center + new Vector3(halfSize.X, halfSize.Y, halfSize.Z);
        var v7 = center + new Vector3(-halfSize.X, halfSize.Y, halfSize.Z);

        scene.AddTriangle(new Triangle(v0, v1, v2, color));
        scene.AddTriangle(new Triangle(v0, v2, v3, color));

        scene.AddTriangle(new Triangle(v5, v4, v7, color));
        scene.AddTriangle(new Triangle(v5, v7, v6, color));

        scene.AddTriangle(new Triangle(v4, v0, v3, color));
        scene.AddTriangle(new Triangle(v4, v3, v7, color));

        scene.AddTriangle(new Triangle(v1, v5, v6, color));
        scene.AddTriangle(new Triangle(v1, v6, v2, color));

        scene.AddTriangle(new Triangle(v3, v2, v6, color));
        scene.AddTriangle(new Triangle(v3, v6, v7, color));

        scene.AddTriangle(new Triangle(v4, v5, v1, color));
        scene.AddTriangle(new Triangle(v4, v1, v0, color));
    }

    private static void CreateFloorPlane(Vector3 center, float size, Color color, Core.Scene scene)
    {
        var halfSize = size / 2f;

        var v0 = center + new Vector3(-halfSize, 0, -halfSize);
        var v1 = center + new Vector3(halfSize, 0, -halfSize);
        var v2 = center + new Vector3(halfSize, 0, halfSize);
        var v3 = center + new Vector3(-halfSize, 0, halfSize);

        scene.AddTriangle(new Triangle(v0, v1, v2, color));
        scene.AddTriangle(new Triangle(v0, v2, v3, color));
    }
}
